### PR TITLE
add multi nodes distributed test support

### PR DESCRIPTION
**Pull request by valuefish**

### CODE REVIEW

**Comment by hellock** [comment]
Thanks for the nice solution!

A concern is that the GPU memory may not be sufficient to perform the gathering if the test dataset is somewhat large. Maybe also provide a CPU method?

**Comment by ZwwWayne** [comment]
I have compared the GPU memory change by `nvidia-smi`, it appears that for the validation set, the GPU memory increases from **1.5G to 1.6G**; when testing RPN on the training set, the GPU memory increases from **1.2G to 10G**.
Therefore, we decide to keep both implementations and make it an option to use GPU or CPU for gathering results.

**Comment by ZwwWayne** [comment]
Sorry for modifying your master branch without your permission because we decide to merge this branch for our rc1 version. 
Thanks for your contribution to the codebase~

File: tools/test.py [hunk 1]
```diff
@@ -100,6 +103,39 @@ def collect_results(result_part, size, tmpdir=None):
...
+
+    if rank == 0:
+        part_list = [
+            pickle.loads(t[:shape_list[i][0]].cpu().numpy().tobytes())
+            for i, t in enumerate(part_recv_list)
+        ]
+        # sort the results
+        ordered_results = []
+        for res in zip(*part_list):
+            ordered_results.extend(list(res))
```
Comment (hellock) [inline tools/test.py]: You may confirm if this is necessary, or simply `mmcv.concat_list(part_list)` is enough.

File: tools/test.py [hunk 2]
```diff
@@ -100,6 +103,39 @@ def collect_results(result_part, size, tmpdir=None):
...
+        part_tensor.new_zeros(shape_max) for _ in range(world_size)
+    ]
+    # gather all result part
+    dist.all_gather(part_recv_list, part_send)
+
+    if rank == 0:
+        part_list = [
+            pickle.loads(t[:shape_list[i][0]].cpu().numpy().tobytes())
+            for i, t in enumerate(part_recv_list)
+        ]
```
Comment (hellock) [inline tools/test.py]: `for recv, shape in zip(part_recv_list, shape_list)` may be clearer.

File: tools/test.py [hunk 3]
```diff
@@ -35,7 +36,7 @@ def single_gpu_test(model, data_loader, show=False):
     return results
 
 
-def multi_gpu_test(model, data_loader, tmpdir=None):
+def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
```
Comment (hellock) [inline tools/test.py]: Add some docstring here since this method is more complex now.

File: tools/test.py (hunk 1 above)
Comment (ZwwWayne) [reply tools/test.py]: This is necessary, it concatenates the list of lists in a different manner from `mmcv.concat_list()`.

File: tools/test.py [hunk 4]
```diff
@@ -35,7 +36,24 @@ def single_gpu_test(model, data_loader, show=False):
...
+    collection. On cpu mode it saves the results on different gpus to 'tmpdir'
+    and collects them by the rank 0 worker.
+    
+    Args:
+        model: Model to be tested.
+        data_loader: Pytorch data loader.
+        tmpdir: Path of directory to save the temporary results from
+            different gpus under cpu mode.
+        gpu_collect: option to use either gpu or cpu for results collection
+    Returns:
```
Comment (hellock) [inline tools/test.py]: Add a blank line before `Returns`.

File: tools/test.py [hunk 5]
```diff
@@ -35,7 +36,24 @@ def single_gpu_test(model, data_loader, show=False):
...
+    """Test model with multiple gpus.
+    
+    This method tests model with multiple gpus and collects the results
+    under two different modes: gpu and cpu modes. By setting 'gpu_collect=True'
+    it encodes results to gpu tensors and use gpu communication for results
+    collection. On cpu mode it saves the results on different gpus to 'tmpdir'
+    and collects them by the rank 0 worker.
+    
+    Args:
+        model: Model to be tested.
```
Comment (hellock) [inline tools/test.py]: arg name (arg type): details and descriptions

File: tools/test.py [hunk 6]
```diff
@@ -35,7 +36,24 @@ def single_gpu_test(model, data_loader, show=False):
...
+        data_loader: Pytorch data loader.
+        tmpdir: Path of directory to save the temporary results from
+            different gpus under cpu mode.
+        gpu_collect: option to use either gpu or cpu for results collection
+    Returns:
+        results: The prediction results.
```
Comment (hellock) [inline tools/test.py]: arg type: details and descriptions

**Comment by openmmlab-v** [comment]
Hi @valuefish ！First of all, we want to express our gratitude for your significant PR in the mmdetection project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/raweFPmdzG

If you have WeChat，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤
